### PR TITLE
feat(http): Switch to pluggable HTTP cache implementation

### PR DESCRIPTION
**Pull request by zharinov**
## Changes

- Create `cacheProvider` option for `Http` class
- Implement `CacheProvider` interface for repository cache
  - Make sure older cache data is compatible with new code
- Refactor `Http` request options type
- Improve tests and coverage

This PR makes it possible to connect package cache with the HTTP caching capabilities.
For example, npm caching logic could be simplified.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/util/http/index.ts [hunk 1]
```diff
@@ -143,13 +118,17 @@ export class Http<Opts extends HttpOptions = HttpOptions> {
...
-      retry: {
-        limit: retryLimit,
-        maxRetryAfter: 0, // Don't rely on `got` retry-after handling, just let it fail and then we'll handle it
+    this.options = merge<GotOptions>(
+      options,
+      {
+        context: { hostType },
+        retry: {
+          limit: retryLimit,
+          maxRetryAfter: 0, // Don't rely on `got` retry-after handling, just let it fail and then we'll handle it
```
Comment (rarkins) [inline lib/util/http/index.ts:127]:
```suggestion
    this.options = merge<GotOptions>(options, {
      context: { hostType },
      retry: {
        limit: retryLimit,
        maxRetryAfter: 0, // Don't rely on `got` retry-after handling, just let it fail and then we'll handle it
```

**Review by rarkins — COMMENTED** [review]
Can this be broken into smaller steps, or too hard?

**Comment by zharinov** [comment]
Yes, let's split

**Review by viceice — COMMENTED** [review]
superceeded 😉

**Comment by zharinov** [comment]
Not really, now we have to switch to the new implementation. I would like to use this PR for this.

**Review by viceice — APPROVED** [review]

**Pull request merged**